rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,296 @@
+  GET_MEDIA: {
+    /** The unique ID of the media */
+    id: string;
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user clicks one of the buttons to copy the media attribution */
+  COPY_ATTRIBUTION: {
+    /** The unique ID of the media */
+    id: string;
+    /** The format of the copied attribution */
+    format: "plain" | "rich" | "html";
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user reports a piece of media through our form */
+  REPORT_MEDIA: {
+    /** The unique ID of the media */
+    id: string;
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user plays, pauses, or seeks an audio track.
+   * @todo: This potentially requires throttling.
+   **/
+  AUDIO_INTERACTION: {
+    /** The unique ID of the media */
+    id: string;
+    event: "seek" | "play" | "pause";
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+  };
+  /** The user visits a CC license description page on CC.org  **/
+  VISIT_LICENSE_PAGE: {
+    /** The slug of the license the user clicked on */
+    license: string;
+  };
+  /** The user visits a creator's link in the single result UI  **/
+  VISIT_CREATOR_LINK: {
+    /** The unique ID of the media */
+    id: string;
+    /** The permalink to the creator's profile */
+    url: string;
+  };
+  /** The user right clicks a single image result, most likely to download it. **/
+  RIGHT_CLICK_IMAGE: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user uses the 'back to search' link on a single result **/
+  BACK_TO_SEARCH: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The visibility of the filter sidebar on desktop is toggled **/
+  TOGGLE_FILTERS: {
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+};
+type AnalyticsEventTypes = keyof AnalyticsEvents;
+
+/**
+ * Send an event to our analytics endpoint. Should be a no-op if the
+ * user has opted-out of analytics.
+ */
+async function sendEvent<T extends AnalyticsEventTypes>(
+  key: T,
+  payload?: AnalyticsEvents[T]
+): Promise<any> {
+  /**
+   * The default values to send with every analytics event.
+   */
+  const defaults = {
+    timestamp: Date.now(),
+    path: window.location.pathname,
+    // The dimensions of the user's device
+    width: window.innerWidth,
+    height: window.innerHeight,
+    referrer: document.referrer || null,
+  };
+  /**
+   * Actually send the event here, with some additional values.
+   * I would have prefered to use `navigator.sendBeacon` here
```

Comment:
I couldn't find any documentation about ad-blockers blocking the `navigator.sendBeacon` API. Rather what ad-blockers seem to be doing is blocking requests to analytics and tracking URLs passed to `sendBeacon`.

Here is empirical evidence with sample size 1.
<img width="355" alt="Screenshot 2022-12-03 at 3 25 26 PM" src="https://user-images.githubusercontent.com/16580576/205438629-7a836ff5-c8a7-4a70-9303-7f1017dbb369.png">

<img width="1000" alt="Screenshot 2022-12-03 at 3 25 38 PM" src="https://user-images.githubusercontent.com/16580576/205438630-3a603ed2-403a-4dd5-8ec1-e4d4a0fe6596.png">

Most ad-blockers will use [EasyList](https://easylist.to/easylist/easylist.txt) so if we avoid endpoints that match the list, we should be able to use `sendBeacon`.

Reply:
uBlock origin blocks it by default, under the default enabled setting "Disable hyperlink auditing":
 
![CleanShot 2022-12-06 at 13 21 42](https://user-images.githubusercontent.com/6351754/205991180-466b9884-9721-484b-b619-81ffecdbdd2e.png)

Details here:

https://github.com/gorhill/uBlock/wiki/Dashboard:-Settings#disable-hyperlink-auditing